File: cpp/buildBeams.py
import numpy as np


def MakeFromPolar(r, theta, phi):
    """
    Convert from polar coordinates to cartesian coordinates
    """
    # get sine and cosine of the angles
    ct = np.cos(theta)
    st = np.sin(theta)
    cp = np.cos(phi)
    sp = np.sin(phi)
    # and convert to cartesian coordinates
    x = r * st * cp
    y = r * st * sp
    z = r * ct
    # create the array
    return np.array([x, y, z])


def buildBeams(N):
    """
    Builds a sphere of unit vectors with N beams in the xy plane.
    """

    # start with an empty list
    beams = []

    # iterate over the whole angle in the xy plane
    for i in range(N):
        # get the Nth part of the total rotation
        theta = (2 * np.pi / N) * i

        # estimate how many vectors we need to cover the phi angle (for the z direction)
        jmax = int(np.floor(N * np.sin(theta) + 0.5))

        # iterate over those angles to get beams in every direction
        for j in range(jmax):
            # get the phi angle
            phi = (2 * np.pi / jmax) * j

            # and create a unit vector from the polar coordinates theta and phi
            beams.append(MakeFromPolar(1.0, theta, phi))

    # return all the vectors
    return np.array(beams)


def saveBeams(beams, fname):
    np.savetxt(fname, beams)

File: saeno/buildBeams.cpp

void buildBeams(std::vector<vec3D>& beams, int N){

    //std::vector<vec3D> beams=std::vector<vec3D>();

    beams.clear();

    double theta,phi;

    vec3D newbeam;

    double pi=3.141592653589793238462643383279502884197169399375105;

    int jmax,j;

    for(int i=0; i<N; i++){

        theta=(2*pi/N)*i;

        jmax=floor(N*sin(theta)+0.5);

        for(j=0; j<jmax; j++){

            phi=(2*pi/jmax)*j;

            newbeam.MakeFromPolar(1.0,theta,phi);

            beams.push_back(newbeam);

        }

    }

    //return beams;

}

void saveBeams(std::vector<vec3D>& beams,const char* fname){

    DRec3D beamsrec=DRec3D();

    beamsrec.data=beams;

    beamsrec.store(fname);

}
